rust/sealevel/client/src/router.rs

```
@@ -555,6 +566,82 @@ fn configure_owner(
+    {
+        programdata_address
+    } else {
+        return;
+    };
+
+    let program_data_account = client.get_account(&programdata_address).unwrap();
+
+    // The program data account must actually a program data account
+    let actual_upgrade_authority = if let Ok(UpgradeableLoaderState::ProgramData {
+        upgrade_authority_address,
+        slot: _,
+    }) = program_data_account.state()
+    {
+        upgrade_authority_address
+    } else {
+        return;
+    };
+
+    let expected_upgrade_authority = Some(router_config.ownable.owner(ctx.payer_pubkey));
+
+    // And the upgrade authority is not what we expect...
+    if actual_upgrade_authority.is_some() && actual_upgrade_authority != expected_upgrade_authority
+    {
+        // Flag if we can't change the upgrade authority
+        if actual_upgrade_authority != Some(ctx.payer_pubkey) {
+            println!(
+                "WARNING: Upgrade authority transfer cannot be completed for chain: {} ({}) from {:?} to {:?}, the existing upgrade authority is not the payer account",
+                chain_config.name,
+                chain_config.domain_id(),
+                actual_upgrade_authority,
+                expected_upgrade_authority,
+            );
+            return;
+        }
+
+        // Then set the upgrade authority to what we expect.
+        ctx.new_txn()
+            .add_with_description(
+                bpf_loader_upgradeable::set_upgrade_authority(
+                    program_id,
+                    actual_upgrade_authority.as_ref().unwrap(),
+                    expected_upgrade_authority.as_ref(),
+                ),
+                format!(
+                    "Setting upgrade authority for chain: {} ({}) to {:?}",
+                    chain_config.name,
+                    chain_config.domain_id(),
+                    expected_upgrade_authority,
+                ),
+            )
+            .with_client(&client)
+            .send_with_payer();
```

Is there a case where this fails silently and we don't end up upgrading? Maybe we can check if the upgrade authority is set correctly after this tx?

---

I'd expect things to fail if it was unsuccessful